File: som-hackathon-starter-dotnet/starters/python/test_producer.py
"""
Publishes seed story.context messages to som.story.context for local testing.

Usage:
    python test_producer.py                     → publishes all seed stories
    python test_producer.py --story informal    → single scenario
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from confluent_kafka import Producer

from config import TOPICS, build_producer_config

SCENARIO_FILES: dict[str, str] = {
    "breaking": "01-breaking-courthouse.json",
    "election": "02-developing-election.json",
    "informal": "03-informal-headline.json",
    "clean": "04-clean-transit.json",
    "breaking-no-compliance": "05-breaking-no-compliance.json",
}

SCENARIOS = list(SCENARIO_FILES.keys())


def _resolve_seed_path(filename: str) -> Path | None:
    """Find a seed story file, trying several candidate locations."""
    candidates = [
        Path.cwd() / ".." / ".." / "seed-stories" / filename,
        Path(__file__).parent / ".." / ".." / "seed-stories" / filename,
        Path.cwd() / "seed-stories" / filename,
    ]
    for c in candidates:
        resolved = c.resolve()
        if resolved.is_file():
            return resolved
    return None


def _delivery_report(err, msg):
    if err:
        print(f"  ✗ Delivery failed: {err}")
    else:
        print(f"  → Partition {msg.partition()}, Offset {msg.offset()}")


def run_test_producer(scenario: str | None = None) -> None:
    producer = Producer(build_producer_config())
    scenarios = [scenario] if scenario else list(SCENARIO_FILES.keys())

    for name in scenarios:
        filename = SCENARIO_FILES.get(name)
        if not filename:
            print(f"Unknown scenario '{name}'. Available: {', '.join(SCENARIO_FILES.keys())}")
            continue

        path = _resolve_seed_path(filename)
        if not path:
            print(f"Seed file not found: {filename}")
            continue

        data = json.loads(path.read_text())
        payload = data.get("payload")
        if not payload:
            print(f"No payload in {filename}")
            continue

        story_id = payload.get("story_id", "unknown")

        print(f"─── {name.upper()} ───")
        print(f"  Story:    {story_id}")
        print(f"  Headline: {payload.get('headline')}")
        print(f"  Phase:    {payload.get('lifecycle', {}).get('phase')}")
        print(f"  File:     {filename}")

        producer.produce(
            TOPICS["story_context"],
            key=story_id,
            value=json.dumps(payload),
            callback=_delivery_report,
        )
        producer.flush()
        print()

    print("Done. Your skill worker should pick these up.")


if __name__ == "__main__":
    args = sys.argv[1:]
    story_idx = args.index("--story") if "--story" in args else -1
    scenario = args[story_idx + 1] if story_idx >= 0 and story_idx + 1 < len(args) else None
    run_test_producer(scenario)

File: som-hackathon-starter-dotnet/starters/node/src/test-producer.ts
/**
 * Publishes seed story.context messages to som.story.context for local testing.
 *
 * Usage:
 *   npx tsx src/index.ts --test-producer                   → publishes all seed stories
 *   npx tsx src/index.ts --test-producer --story informal  → single scenario
 */

import { readFileSync, existsSync } from "fs";
import path from "path";
import { Kafka } from "kafkajs";
import { buildKafkaConfig, TOPICS } from "./config.js";

const SCENARIO_FILES: Record<string, string> = {
  breaking: "01-breaking-courthouse.json",
  election: "02-developing-election.json",
  informal: "03-informal-headline.json",
  clean: "04-clean-transit.json",
  "breaking-no-compliance": "05-breaking-no-compliance.json",
};

export const SCENARIOS = Object.keys(SCENARIO_FILES);

function resolveSeedPath(filename: string): string | null {
  // Try ../../seed-stories/ (relative to project root, shared with .NET)
  const candidates = [
    path.join(process.cwd(), "..", "..", "seed-stories", filename),
    path.join(__dirname, "..", "..", "..", "seed-stories", filename),
    path.join(process.cwd(), "seed-stories", filename),
  ];
  for (const c of candidates) {
    if (existsSync(c)) return c;
  }
  return null;
}

export async function runTestProducer(scenario?: string): Promise<void> {
  const kafka = new Kafka(buildKafkaConfig("som-test-producer"));
  const producer = kafka.producer();
  await producer.connect();

  const scenarios = scenario ? [scenario] : Object.keys(SCENARIO_FILES);

  for (const name of scenarios) {
    const file = SCENARIO_FILES[name];
    if (!file) {
      console.log(
        `Unknown scenario '${name}'. Available: ${Object.keys(SCENARIO_FILES).join(", ")}`
      );
      continue;
    }

    const fullPath = resolveSeedPath(file);
    if (!fullPath) {
      console.log(`Seed file not found: ${file}`);
      continue;
    }

    const raw = readFileSync(fullPath, "utf-8");
    const envelope = JSON.parse(raw);
    const payload = envelope.payload;
    if (!payload) {
      console.log(`No payload in ${file}`);
      continue;
    }

    const storyId: string = payload.story_id ?? "unknown";

    console.log(`─── ${name.toUpperCase()} ───`);
    console.log(`  Story:    ${storyId}`);
    console.log(`  Headline: ${payload.headline}`);
    console.log(`  Phase:    ${payload.lifecycle?.phase}`);
    console.log(`  File:     ${file}`);

    const result = await producer.send({
      topic: TOPICS.storyContext,
      messages: [{ key: storyId, value: JSON.stringify(payload) }],
    });

    const record = result[0];
    console.log(
      `  → Partition ${record.partition}, Offset ${record.baseOffset}`
    );
    console.log();
  }

  await producer.disconnect();
  console.log("Done. Your skill worker should pick these up.");
}
